plugins/deisa/deisa.cxx

```diff
@@ -0,0 +1,206 @@
+#include <pdi/datatype.h>
+#include <pdi/expression.h>
+#include <pdi/logger.h>
+#include <pdi/paraconf_wrapper.h>
+#include <pdi/plugin.h>
+#include <pdi/python/tools.h>
+#include <pdi/ref_any.h>
+#include <pdi/scalar_datatype.h>
+
+namespace {
+
+namespace py = pybind11;
+
+using namespace PDI;
+using namespace py::literals;
+
+/** The deisa plugin
+ */
+class deisa_plugin: public Plugin
+{
+	static constexpr char DEISA_COMPATIBLE_VERSION[] = "0.3.2"; // Used to check compatibility with Deisa's python library
+
+	Expression m_scheduler_info;
+	std::unordered_map<std::string, Datatype_template_sptr> m_deisa_arrays;
+	Expression m_time_step;
+	py::object m_bridge = py::none();
+
+public:
+	static std::pair<std::unordered_set<std::string>, std::unordered_set<std::string>> dependencies() { return {{"mpi"}, {"mpi"}}; }
+
+	deisa_plugin(Context& ctx, PC_tree_t conf)
+		: Plugin{ctx}
+	{
+		if (!Py_IsInitialized()) {
+			py::initialize_interpreter();
+		}
+
+		// init params
+		each(conf, [&](PC_tree_t key_tree, PC_tree_t value) {
+			std::string key = to_string(key_tree);
+			if (key == "scheduler_info") {
+				m_scheduler_info = to_string(value);
+			} else if (key == "deisa_arrays") {
+				each(value, [&](PC_tree_t key_map, PC_tree_t value_map) { m_deisa_arrays.emplace(to_string(key_map), ctx.datatype(value_map)); });
+			} else if (key == "time_step") {
+				m_time_step = to_string(value);
+			} else if (key == "map_in") {
+				//
+			} else if (key == "logging" || key == "init_on") {
+				//
+			} else {
+				throw Config_error{key_tree, "Unknown key in Deisa file configuration: '{}'", key};
+			}
+		});
+
+		// plugin init
+		PC_tree_t init_tree = PC_get(conf, ".init_on");
+		if (!PC_status(init_tree)) {
+			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(ctx); }, to_string(init_tree));
+		} else {
+			throw Config_error{conf, "Deisa plugin requires init_on key "};
+		} // TODO: replace with try/catch when #480 is fixed.
+
+		// map_in
+		PC_tree_t map_in = PC_get(conf, ".map_in");
+		if (!PC_status(map_in)) {
+			each(map_in, [&](PC_tree_t key_map, PC_tree_t value_map) { // TODO: when #481 is fixed, use `or_none` variant.
+				ctx.callbacks().add_data_callback(
+					[&, deisa_array_name = to_string(value_map)](const std::string&, const Ref& data_ref) {
+						try {
+							assert(m_bridge != py::none());
+							assert(hasattr(m_bridge, "publish_data"));
+							py::object publish_data = m_bridge.attr("publish_data");
+#ifdef NDEBUG
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a = false);
+#else
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a=true);
+#endif
+						} catch (const std::exception& e) {
+							throw Plugin_error("While publishing data. Caught exception: {}", std::string(e.what()));
+						} catch (...) {
+							throw Plugin_error("While publishing data.");
+						}
+					},
```

Comment:
What would you say about moving this to a dedicated `on_data(const std::string&, const Ref& data_ref)` member function?